pinot-server/src/main/java/com/linkedin/pinot/server/starter/helix/SegmentOnlineOfflineStateModelFactory.java

```diff
@@ -89,26 +92,51 @@ public SegmentOnlineOfflineStateModel(String helixClusterName, String instanceId
 
     @Transition(from = "OFFLINE", to = "CONSUMING")
     public void onBecomeConsumingFromOnline(Message message, NotificationContext context) {
-      LOGGER.error("Unexpected transition from OFFLINE to CONSUMING for {}", message.getResourceName());
-      throw new RuntimeException("Unexpected state transition");
+      LOGGER.info("SegmentOnlineOfflineStateModel.onBecomeConsumingFromOffline() : " + message);
+      // We do the same processing as usual for going to the consuming state, which adds the segment to the table data
+      // manager and starts Kafka consumption
+      onBecomeOnlineFromOffline(message, context);
     }
 
     @Transition(from = "CONSUMING", to = "ONLINE")
     public void onBecomeOnlineFromConsuming(Message message, NotificationContext context) {
-      LOGGER.error("Unexpected transition from CONSUMING to ONLINE for {}", message.getResourceName());
-      throw new RuntimeException("Unexpected state transition");
+      final TableDataManager tableDataManager =
+          ((InstanceDataManager) INSTANCE_DATA_MANAGER).getTableDataManager(message.getResourceName());
+      SegmentDataManager acquiredSegment = tableDataManager.acquireSegment(message.getPartitionName());
```

Comment:
acquireSegment takes a read-lock on the segment. we don't want to hold a readlock for a long time (go online from consuming can take a while). Also, if at the end of it, we want to replace the segment, we need to take the write lock, so this won't work.